core/src/main/java/org/jboss/pnc/build/finder/pnc/client/CachedPncClient.java

```diff
@@ -18,77 +18,103 @@
 
-    private final Map<String, StaticRemoteCollection<Artifact>> sha256Cache = new HashMap<>();
+    private final Map<String, ArtifactStaticRemoteCollection> artifactCache;
 
     private final Map<String, BuildPushResult> getBuildPushResultCache = new HashMap<>();
 
     private final Map<String, ProductVersion> getProductVersionCache = new HashMap<>();
 
-    public HashMapCachingPncClient(BuildConfig config) {
+    public CachedPncClient(BuildConfig config, BasicCacheContainer cacheManager) {
+        if (cacheManager == null) {
+            artifactCache = new HashMap<>();
+        } else {
+            artifactCache = cacheManager.getCache("artifact-pnc");
+        }
         this.pncClient = new PncClientImpl(config);
     }
 
-    public HashMapCachingPncClient(PncClient pncClient) {
+    public CachedPncClient(PncClient pncClient, BasicCacheContainer cacheManager) {
+        if (cacheManager == null) {
+            artifactCache = new HashMap<>();
+        } else {
+            artifactCache = cacheManager.getCache("artifact-pnc");
+        }
         this.pncClient = pncClient;
     }
 
     @Override
     public RemoteCollection<Artifact> getArtifactsByMd5(String md5) throws RemoteResourceException {
-        RemoteCollection<Artifact> cachedEntity = md5Cache.get(md5);
-        if (cachedEntity != null) {
-            return cachedEntity;
-        } else {
-            RemoteCollection<Artifact> foundEntity = pncClient.getArtifactsByMd5(md5);
-            StaticRemoteCollection<Artifact> staticRemoteCollection = new StaticRemoteCollection<>(foundEntity);
-            md5Cache.put(md5, staticRemoteCollection);
-            return staticRemoteCollection;
+        ArtifactStaticRemoteCollection cachedValue = getFromCache(md5);
+        if (cachedValue != null) {
+            return cachedValue;
+        }
+
+        RemoteCollection<Artifact> artifacts = pncClient.getArtifactsByMd5(md5);
+        if (artifacts != null && artifacts.size() > 0) {
+            insertToCache(md5, artifacts);
         }
+
+        return artifacts;
     }
 
     @Override
     public RemoteCollection<Artifact> getArtifactsBySha1(String sha1) throws RemoteResourceException {
-        RemoteCollection<Artifact> cachedEntity = sha1Cache.get(sha1);
-        if (cachedEntity != null) {
-            return cachedEntity;
-        } else {
-            RemoteCollection<Artifact> foundEntity = pncClient.getArtifactsBySha1(sha1);
-            StaticRemoteCollection<Artifact> staticRemoteCollection = new StaticRemoteCollection<>(foundEntity);
-            sha1Cache.put(sha1, staticRemoteCollection);
-            return staticRemoteCollection;
+        ArtifactStaticRemoteCollection cachedValue = getFromCache(sha1);
+        if (cachedValue != null) {
+            return cachedValue;
+        }
+
+        RemoteCollection<Artifact> artifacts = pncClient.getArtifactsBySha1(sha1);
+        if (artifacts != null && artifacts.size() > 0) {
```

Comment:
Do we not want to cache empty? It is a valid result?

Reply:
While the empty result may be valid, the artifact could appear in PNC or Brew after some time. If we cached the empty result and the artifact was imported, we would never know about this because the empty result would be returned even between BF runs until the ISPN cache was dropped.

Reply:
The local cache used by the CLI caches empty. If it doesn't expire entries, then that is bad. But, if entries expire at some point, it should be OK.

I am OK with not caching empty except that we would guess that most hits return "not found" and this is where we'd get the most benefit from caching.

Reply:
I wouldn't cache for now and add a cache expiry with caching later in next issue.